### PR TITLE
Create rofication block

### PR DESCRIPTION
**Pull request by cfsmp3**
Rofication is the notification system used by regolith linux (and others I assume, but this is my use case).

It's a simple system - there's a daemon that receives the notications on d-bus and that exports them via a UNIX socket. The block itself just needs to display a counter and a color (based on any notification being critical or not) and spawn the GUI when clicked.

### CODE REVIEW

**Comment by cfsmp3** [comment]
@greshake Could you review?

**Comment by ammgws** [comment]
Does this support both the [original](https://github.com/DaveDavenport/Rofication) rofication as well as the regolith fork? If not can you mention that in the docs?

File: src/blocks/rofication.rs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    text: TextWidget,
+    update_interval: Duration,
+
+    //useful, but optional
+    #[allow(dead_code)]
+    shared_config: SharedConfig,
+    #[allow(dead_code)]
+    tx_update_request: Sender<Task>,
+    // UNIX socket to read from
+    pub socket_path: String,
```
Comment (ammgws) [inline src/blocks/rofication.rs:41]: Can you add a `format` string too? It seems inevitable so we may as well do it from the start.

File: src/blocks/rofication.rs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+    pub interval: Duration,
+    // UNIX socket to read from
+    pub socket_path: String,
+}
+
+impl Default for RoficationConfig {
+    fn default() -> Self {
+        Self {
+            interval: Duration::from_secs(1),
+            socket_path: String::from_str("/tmp/rofi_notification_daemon").unwrap(),
```
Comment (ammgws) [inline src/blocks/rofication.rs:58]: Why not just `"/tmp/rofi_notification_daemon".to_string()` or `String::from("/tmp/rofi_notification_daemon".to_string()")`?

File: src/blocks/rofication.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    type Config = RoficationConfig;
+
+    fn new(
+        id: usize,
+        block_config: Self::Config,
+        shared_config: SharedConfig,
+        tx_update_request: Sender<Task>,
+    ) -> Result<Self> {
+        let text = TextWidget::new(id, 0, shared_config.clone())
+            .with_text("0")
```
Comment (ammgws) [inline src/blocks/rofication.rs:73]: Should this just be a blank string `""` or `"?"` instead since we don't have the actual count yet?

**Review by ammgws — CHANGES_REQUESTED** [review]

**Comment by cfsmp3** [comment]
> Does this support both the [original](https://github.com/DaveDavenport/Rofication) rofication as well as the regolith fork? If not can you mention that in the docs?

Yes, it only uses one command to get the number of notifications and that came from upstream.

I personally use, and contribute to, the fork but it is a superset of the original.

File: src/blocks/rofication.rs (hunk 1 above)
Comment (cfsmp3) [reply src/blocks/rofication.rs:39]: > Can you add a format string too? It seems inevitable so we may as well do it from the start.

Done.

File: src/blocks/rofication.rs (hunk 2 above)
Comment (cfsmp3) [reply src/blocks/rofication.rs]: Changed.

File: src/blocks/rofication.rs (hunk 3 above)
Comment (cfsmp3) [reply src/blocks/rofication.rs]: Changed to ?.

**Review by ammgws — APPROVED** [review]

**Comment by ammgws** [comment]
Thanks!

**Comment by MaxVerevkin** [comment]
> > Does this support both the [original](https://github.com/DaveDavenport/Rofication) rofication as well as the regolith fork? If not can you mention that in the docs?
> 
> Yes, it only uses one command to get the number of notifications and that came from upstream.
> 
> I personally use, and contribute to, the fork but it is a superset of the original.

Actually, it doesn't work with the original.

- It works only with if the command is `b"num"`, not `b"num\n"`.
- The original separates `num` and `critical` with a new line, not a comma.

**Comment by MaxVerevkin** [comment]
So, perhaps rename it?

**Comment by MaxVerevkin** [comment]
Or, if the fork forks with `"b"num"` too, then the block can be adapted to work with both versions.

**Comment by cfsmp3** [comment]
Nah, let's do things right. I'm away but I'll fix this properly asap.

**Comment by cfsmp3** [comment]
@MaxVerevkin https://github.com/greshake/i3status-rust/pull/1390

**Pull request merged**